sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2461,6 +2519,174 @@ checkDevSupportDeviceRequirements(const device &Dev,
+    std::string SingleJointMatrixMad;
+
+    std::vector<ext::oneapi::experimental::matrix::combination> Combinations =
+        Dev.get_info<
+            ext::oneapi::experimental::info::device::matrix_combinations>();
+
+    // start to parse the value which is generated by
+    // SYCLPropagateJointMatrixUsage pass
+    while (std::getline(JointMatrixMadStrStream, SingleJointMatrixMad, ';')) {
+      std::istringstream SingleJointMatrixMadStrStream(SingleJointMatrixMad);
+      std::vector<std::string> JointMatrixMadVec;
+      std::string Item;
+
+      while (std::getline(SingleJointMatrixMadStrStream, Item, ',')) {
+        JointMatrixMadVec.push_back(Item);
+      }
+
+      if (JointMatrixMadVec.size() == 7) {
+        std::string MatrixTypeAStrUser = JointMatrixMadVec[0];
+        std::string MatrixTypeBStrUser = JointMatrixMadVec[1];
+        std::string MatrixTypeCStrUser = JointMatrixMadVec[2];
+        std::string MatrixTypeDStrUser = JointMatrixMadVec[3];
+        size_t MSizeUser = std::stoi(JointMatrixMadVec[4]);
+        size_t KSizeUser = std::stoi(JointMatrixMadVec[5]);
+        size_t NSizeUser = std::stoi(JointMatrixMadVec[6]);
+
+        std::optional<matrix_ext::matrix_type> MatrixTypeAUserOpt =
+            convertMatrixTypeStringMatrixTypeEnumValue(MatrixTypeAStrUser);
+        std::optional<matrix_ext::matrix_type> MatrixTypeBUserOpt =
+            convertMatrixTypeStringMatrixTypeEnumValue(MatrixTypeBStrUser);
+        std::optional<matrix_ext::matrix_type> MatrixTypeCUserOpt =
+            convertMatrixTypeStringMatrixTypeEnumValue(MatrixTypeCStrUser);
+        std::optional<matrix_ext::matrix_type> MatrixTypeDUserOpt =
+            convertMatrixTypeStringMatrixTypeEnumValue(MatrixTypeDStrUser);
+
+        bool IsMatrixMadCompatible = false;
+
+        for (const auto &Combination : Combinations) {
+          if (MatrixTypeAUserOpt && MatrixTypeBUserOpt && MatrixTypeCUserOpt &&
+              MatrixTypeDUserOpt) {
+            IsMatrixMadCompatible =
+                (MatrixTypeAUserOpt.value() == Combination.atype) &&
+                (MatrixTypeBUserOpt.value() == Combination.btype) &&
+                (MatrixTypeCUserOpt.value() == Combination.ctype) &&
+                (MatrixTypeDUserOpt.value() == Combination.dtype) &&
+                ((Combination.msize != 0)
+                     ? (MSizeUser == Combination.msize)
+                     : (MSizeUser <= Combination.max_msize)) &&
+                ((Combination.ksize != 0)
+                     ? (KSizeUser == Combination.ksize)
+                     : (KSizeUser <= Combination.max_ksize)) &&
+                ((Combination.nsize != 0)
+                     ? (NSizeUser == Combination.nsize)
+                     : (NSizeUser <= Combination.max_nsize));
+          }
+
+          // early exit if we have a match
+          if (IsMatrixMadCompatible)
+            break;
+        }
+
+        if (!IsMatrixMadCompatible)
+          return sycl::exception(
```

Comment:
@intel/sycl-matrix-reviewers, 3 tests below failed on Intel Arc A-Series Graphics. Looks like they use incompatible matrix parameters. Should they use N == 16, and not == 8 as it is now?

```
Failed Tests (3):
  SYCL :: Matrix/XMX8/element_wise_ops.cpp
  SYCL :: Matrix/XMX8/joint_matrix_all_sizes.cpp
  SYCL :: Matrix/XMX8/joint_matrix_ss_int8.cpp
```

Reply:
Also these tests below failed on CUDA. Looks like Rows or Cols is incorrect, or both. Please suggest the correct numbers. 
Please also note that you will likely get similar failures for other architectures after the internal QA test run.

```
Failed Tests (3):
  SYCL :: Matrix/joint_matrix_tensorcores_sm70.cpp
  SYCL :: Matrix/joint_matrix_tensorcores_sm72.cpp
  SYCL :: Matrix/joint_matrix_tensorcores_sm80.cpp
```

Reply:
> @intel/sycl-matrix-reviewers, 3 tests below failed on Intel Arc A-Series Graphics. Looks like they use incompatible matrix parameters. Should they use N == 16, and not == 8 as it is now?
> 
> ```
> Failed Tests (3):
>   SYCL :: Matrix/XMX8/element_wise_ops.cpp
>   SYCL :: Matrix/XMX8/joint_matrix_all_sizes.cpp
>   SYCL :: Matrix/XMX8/joint_matrix_ss_int8.cpp
> ```

@dm-vodopyanov, No, N=8 for DG2.
This may be a bug in the runtime query?

Reply:
> Also these tests below failed on CUDA. Looks like Rows or Cols is incorrect, or both. Please suggest the correct numbers. Please also note that you will likely get similar failures for other architectures after the internal QA test run.
> 
> ```
> Failed Tests (3):
>   SYCL :: Matrix/joint_matrix_tensorcores_sm70.cpp
>   SYCL :: Matrix/joint_matrix_tensorcores_sm72.cpp
>   SYCL :: Matrix/joint_matrix_tensorcores_sm80.cpp
> ```

@dm-vodopyanov, the runtime query does not provide information about Nvidia usage. So these tests should not be affected. Remember, the Nvidia usage is only supported in AOT compilation mode.

Reply:
Thanks @dkhaldi, I will debug these two cases

Reply:
Fixed on DG2, there was a typo in `matrix_combinations`:
```
Failed Tests (3):
  SYCL :: Matrix/XMX8/element_wise_ops.cpp
  SYCL :: Matrix/XMX8/joint_matrix_all_sizes.cpp
  SYCL :: Matrix/XMX8/joint_matrix_ss_int8.cpp
```

These tests on CUDA fail due to runtime query is not yet implemented on CUDA ([runtime query for CUDA in the spec](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc#nvidia-tensor-cores-supported-combinations)): 
```
Failed Tests (3):
  SYCL :: Matrix/joint_matrix_tensorcores_sm70.cpp
  SYCL :: Matrix/joint_matrix_tensorcores_sm72.cpp
  SYCL :: Matrix/joint_matrix_tensorcores_sm80.cpp
```
@dkhaldi what should we do in two following cases?

1. If device is CUDA, should we ignore optional kernel feature for `joint_matrix` and `joint_matrix_mad` until runtime query for CUDA is implemented?
2. (not related to these tests) To double-check, if user tries to use `joint_matrix` and `joint_matrix_mad` on incompatible device, let's say on Coffelake, should we throw the exception implemented in this patch?

Reply:
@JackAKirk, what is Codeplay plans on adding the runtime query? Once this is added, @dm-vodopyanov work here to add joint_matrix as an optional kernel feature should just work for the Nvidia case.

In the meantime, I think this should be ignored so we should avoid having these tests fail.

Reply:
CC @npmiller for deciding roadmap priority of this.

Reply:
CUDA and HIP are now ignored, until `matrix_combinations` for them is implemented. Added a TODO about that.